ui/hooks/useAsyncResult.ts

```diff
@@ -0,0 +1,75 @@
+import { useState, DependencyList, useEffect } from 'react';
+
+/**
+ * Represents the result of an asynchronous function where errors
+ * are thrown to be handled by an error boundary.
+ */
+export type AsyncResultStrict<T> =
+  | { pending: true; value?: never } // pending
+  | { pending: false; value: T }; // success
+
+/**
+ * Represents the result of an asynchronous function with the
+ * possibility of an error
+ */
+export type AsyncResult<T> =
+  | (AsyncResultStrict<T> & { error?: never })
+  | { pending: false; value?: never; error: Error }; // error
+
+/**
+ * Hook that executes an asynchronous function and returns its result
+ * or an error (errors are caught and returned as part of the result).
+ *
+ * @param asyncFn
+ * @param dependencies
+ */
+export function useAsyncResult<T>(
+  asyncFn: () => Promise<T>,
+  dependencies: DependencyList = [],
+) {
+  const [result, setResult] = useState<AsyncResult<T>>({
+    pending: true,
+  });
+
+  useEffect(() => {
+    setResult({ pending: true });
+    let cancelled = false;
+    asyncFn()
+      .then((value) => {
+        if (!cancelled) {
+          setResult({ pending: false, value });
+        }
+      })
+      .catch((error) => {
+        if (!cancelled) {
+          setResult({ pending: false, error: error as Error });
+        }
+      });
+    return () => {
+      cancelled = true;
+    };
+  }, dependencies);
+
+  return result;
+}
+
+/**
+ * Hook that executes an asynchronous function and returns its result
+ * or throws an error to be handled by an error boundary.
+ *
+ * @param asyncFn
+ * @param deps
+ * @returns
+ */
+export function useAsyncResultStrict<T>(
```

Comment:
When I read `Strict` I think of value comparison or typing.

Would it more explicit if we called it something like `useAsyncResultOrThrow`?